Removes LeaveMap window.

deniz1a opened the pull request:
So that it is possible to look at the map after game has ended. The map is frozen, it is possible to select units but not order them. I added buttons to ra, td and d2k mods.

deniz1a commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/LeaveMapLogic.cs:
```diff
@@ -97,6 +97,16 @@ public LeaveMapLogic(Widget widget, World world, OrderManager orderManager)
+				if (mpe != null)
+					mpe.Fade(mpe.Info.Effect);
+
+				Game.LoadWidget(world, "INGAME_ROOT", widget, new WidgetArgs());
```
I'm not sure if passing the widget variable here is correct, but it works.

obrakmann commented:
This works nicely overall. There is one problem though: if the map defines a win/lose video, it plays every time the dialog appears, not just once. I'm not sure if another approach that would just hide the endgame widget would be better.

deniz1a commented:
Here are some screenshots:
Red Alert
Full
![rtm1](https://cloud.githubusercontent.com/assets/6599106/7526621/ecabfe08-f51b-11e4-8a9b-d30714fcc04f.png)

Simple
![rtm2](https://cloud.githubusercontent.com/assets/6599106/7526625/f7af28f2-f51b-11e4-92f4-2b17eca01721.png)

Tiberian Dawn
Full
![rtm3](https://cloud.githubusercontent.com/assets/6599106/7526638/12966ec8-f51c-11e4-9dbd-a331dee0d29a.png)

Simple
![rtm4](https://cloud.githubusercontent.com/assets/6599106/7526642/1e0eb9e0-f51c-11e4-9f3e-afb56466f3eb.png)

Dune 2000
Full
![rtm5](https://cloud.githubusercontent.com/assets/6599106/7526649/2a929c90-f51c-11e4-9273-c5ebbd430d53.png)

deniz1a commented:
OK I'll try to just hide the widget and not reload it every time.

deniz1a commented:
How do I close the leave map window? I tried Ui.CloseWindow() but it doesn't work. Also the options button on the side panel is more complicated... Would it be easier to change the video logic? I can't find its code.

Does the current implementation have any disadvantages other than multiple video loading?

deniz1a commented:
Which mission or custom map has game end video? So I can test it.

deniz1a commented:
Yeah actually the way I had implemented this was completely wrong. I was reloading INGAME_ROOT everytime the window was closed. I changed that now and it acts like ingame menu.

deniz1a commented:
This patch also makes the chat text field automatically get keyboard focus when chat tab is clicked.

deniz1a commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/LoadIngamePlayerOrObserverUILogic.cs:
```diff
@@ -45,9 +44,13 @@ public LoadIngamePlayerOrObserverUILogic(Widget widget, World world)
+				if (w == null)
+					w = Ui.Root.GetOrNull("PLAYER_WIDGETS");
+
+				var optionsButton = w == null ? null : w.GetOrNull<MenuButtonWidget>("OPTIONS_BUTTON");
```
Do I need a null check here? It's either observer widgets or player widgets... The ingame menu code had null check for options button so I put it here below too. But is there a side panel without options button?

Phrohdoh replied:
You assign to `w` above if it is `null`.

deniz1a replied:
But if w can be null then I have to check for that in line 51. Because null.GetOrNull() is an exception?

Phrohdoh replied:
Try it and see.

pchote replied:
@Phrohdoh: If neither `OBSERVER_WIDGETS` or `PLAYER_WIDGETS` exist then it will be null at this line.

deniz1a replied:
Yes, but is it possible that neither of those exist at GameOver?

deniz1a replied:
OK, checking if it's null wouldn't affect performance in any noticable way I presume. So it's ok to leave it as it is...

Phrohdoh replied:
@pchote Yes I know.

deniz1a commented:
I think this PR should wait for #8118 to be merged. It changes some keyboard focus behavior and I may need to adjust this accordingly. Some yield focus method calls may become unnecessary then.

deniz1a commented:
Has dependencies label can be removed, I closed #8118. Also, this PR is not yet finished, there are some glitches, I'm trying to fix them.

deniz1a commented:
Should I merge leave map widget into ingame menu? So it would be a version of the ingame menu that is displayed after game ends?

deniz1a commented:
Leave map window could have settings and music buttons too...

deniz1a committed:
Freezes map after game ends.

Adds newline.

Disables keyboard hotkeys for units after game ends.

deniz1a committed:
Merges LeaveMapLogic into GameInfoLogic.

Opens options menu when game ends.

Closes settings or music window before opening options menu.

Moves game end video playback logic from GameInfoLogic to
LoadIngamePlayerOrObserverUILogic.

Improves menu buttons.

Stop video from playing on fatal lua script error

deniz1a commented:
OK, I think this feature is finished. This patch makes it possible to look at the map after game has ended. Map is frozen, it is only possible to select units. I merged the script error and video display parts of LeaveMapLogic into GameInfoLogic, so they are shown in options menu now. Also removed all code related to leave map window.

Needs review and testing.

Phrohdoh commented on OpenRA.Game/Selection.cs:
```diff
@@ -57,6 +57,15 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
 
+			foreach (var a in newSelection)
+				foreach (var sel in a.TraitsImplementing<INotifySelected>())
+					sel.Selected(a);
```
Newline after this.

Phrohdoh commented on OpenRA.Game/Selection.cs:
```diff
@@ -57,6 +57,15 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
+			foreach (var ns in world.WorldActor.TraitsImplementing<INotifySelection>())
+				ns.SelectionChanged();
+
+			if (world.IsGameOver)
```
Shouldn't this be done before the above `foreach`s?

deniz1a replied:
I don't know. I only wanted to prevent unit voices from being played. I didn't want to change anything else about selections.

GraionDilach commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoLogic.cs:
```diff
@@ -38,13 +50,22 @@ public GameInfoLogic(Widget widget, World world, IngameInfoPanel activePanel)
+
+				Game.BeforeGameStart += ResetVideoPlayed;
+				string video = null;
+				if (world.LocalPlayer.WinState != WinState.Undefined)
```
This will play the video immediately when the player win/lose, breaking cooperative missions/multiplayer - teams can still win after an individual player has lost the game.

deniz1a replied:
I copied that from current Leave Map widget code. How should it be changed?

GraionDilach replied:
This should be moved out to a separate function triggered by the GameOver Action, for example.

deniz1a replied:
Oh ok, so it should be played when game is over, not when an individual player loses...

deniz1a commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoLogic.cs:
```diff
@@ -38,13 +43,23 @@ public GameInfoLogic(Widget widget, World world, IngameInfoPanel activePanel)
+						Media.PlayFMVFullscreen(world, video, () => { });
+				};
+				world.GameOver -= playEndVideo;
+				world.GameOver += playEndVideo;
```
Why is this not working?

penev92 commented:
So this removes the `LEAVE_MAP_WIDGET`, moves the logic to the ingame menu, and uses that when the game ends, is that correct?

Also adds a `Return to map` button to that so you can spectate after the game is over.

GraionDilach commented:
I guess what it does now is that it freezes the game instead of throwing the `LEAVE_MAP_WIDGET` so you have to manually quit. I don't really understand the intention behind this tbh. Pretty sure players would mostly quit immediately and only 5% of the matches would end up with people checking out map after the game.

deniz1a commented:
This PR does the following:
* Removes LeaveMap window.
* Doesn't show order cursor (move, attack etc.) after game end
* Doesn't play unit selection voices after game end
* Opens options menu when game ends.

That's it. It doesn't add a ```Return to map``` button to options menu because that already has ```Resume``` button. So after game ends, you get the options menu and you can close it and look at the map.

The reason for this patch is to be able to look at the map after game end. There is no difference in how you quit the game. Instead of pressing ```Leave``` button in LeaveMap window, you click ``Àbort Mission``` button in options menu.

GraionDilach commented:
You removed endgame statistics, congratulations! :-1: for removing the widget.

penev92 commented:
This PR started good, and I think people were generally happy with it. It has changed quite a lot since then though, and we have to reevaluate that.
Here is my opinion:
 - Having to press `Abort Mission` at that point is just... confusing. Also is `Surrender` still active?
 - The current setup has statistics and stuff which we don't want removed.
 - Having a dedicated endgame widget allows mods to do all kinds of stuff at the end of a game/mission, like playing videos in a window, having cool banners, etc., which is not possible when we remove it and use the ingame menu.
So I'll have to :-1: the current implementation and request that you go back to a `Return to map` button.

deniz1a commented:
Rebased.

When game ends, observer widgets are loaded. Statistics window can be opened with the button next to the one for the options menu.

Should the text of Abort mission button be changed to Leave after game ends? Surrender button is disabled for observers.

Maps can include custom game end video. Other than that, those customizations could be implemented for options menu too...

penev92 commented:
- It took me 2 games until I figured that I need to press "Resume" to go back and look at the map, which is something that will need to be discussed.
 - User keyboard input is not ignored - you can order actors to deploy/undeploy, guard, attack, etc.

Leaving the `Changes requested` label for the second point.

deniz1a commented:
Ok, I'll disable keyboard orders and change Resume button to Return to map.

deniz1a commented:
I disabled unit hotkeys, changed ```Resume``` to ```Return to map``` and ```Abort mission``` to ```Leave```. Also disabled all menu buttons and debug tab button. Needs review.

deniz1a commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/LoadIngamePlayerOrObserverUILogic.cs:
```diff
@@ -46,9 +47,19 @@ public LoadIngamePlayerOrObserverUILogic(Widget widget, World world)
+				{
+					var video = world.LocalPlayer.WinState == WinState.Won ? world.Map.Videos.GameWon : world.Map.Videos.GameLost;
+					if (!string.IsNullOrEmpty(video))
+						Media.PlayFMVFullscreen(world, video, () => { });
```
I think this could also be done in Media if it implemented IGameOver interface. Should I move this to Media.GameOver()?

obrakmann commented:
Please remove `World.ObserveAfterWinOrLose` too, then. That was added to prevent the the observer UI from showing for a split second in missions before the video starts, and isn't useful anymore with this. Ideally, loading of the observer UI should be delayed by the `GameOverDelay` defined in `MissionObjectivesInfo` (see diff below), but that causes some weirdness in the radar widget, of all things.

When a lua script errors, the video is still shown before the error dialog pops up.

I'm not really sure if this isn't ultimately a step backwards wrt to what we'll eventually want to do with the score screen, but since we don't have a clearly laid out plan for that, and this will please the score screen haters, I guess I'm overall ok with the way this works. I'm not that fond of the many `IsGameOver` checks sprinkled throughout a load of unrelated traits, though, but don't see any better way either atm, and in principle it's not much different to the `player.WinState` checks found everywhere.

```diff
diff --git a/OpenRA.Game/World.cs b/OpenRA.Game/World.cs
index daabba3..e1c9a92 100644
--- a/OpenRA.Game/World.cs
+++ b/OpenRA.Game/World.cs
@@ -66,11 +66,10 @@ public void EndGame()
 			}
 		}
 
-		public bool ObserveAfterWinOrLose;
 		Player renderPlayer;
 		public Player RenderPlayer
 		{
-			get { return renderPlayer == null || (ObserveAfterWinOrLose && renderPlayer.WinState != WinState.Undefined) ? null : renderPlayer; }
+			get { return renderPlayer == null || renderPlayer.WinState != WinState.Undefined ? null : renderPlayer; }
 			set { renderPlayer = value; }
 		}
 
diff --git a/OpenRA.Mods.Common/Traits/Player/MissionObjectives.cs b/OpenRA.Mods.Common/Traits/Player/MissionObjectives.cs
index 8e80825..e1c64d5 100644
--- a/OpenRA.Mods.Common/Traits/Player/MissionObjectives.cs
+++ b/OpenRA.Mods.Common/Traits/Player/MissionObjectives.cs
@@ -53,7 +53,7 @@ public class MissionObjectivesInfo : ITraitInfo
 
 	public class MissionObjectives : INotifyObjectivesUpdated, ISync, IResolveOrder
 	{
-		readonly MissionObjectivesInfo info;
+		public readonly MissionObjectivesInfo Info;
 		readonly List<MissionObjective> objectives = new List<MissionObjective>();
 		public ReadOnlyList<MissionObjective> Objectives;
 
@@ -64,12 +64,10 @@ public class MissionObjectives : INotifyObjectivesUpdated, ISync, IResolveOrder
 		// The player's WinState is only updated when his allies have all completed their objective as well.
 		public WinState WinStateCooperative { get; private set; }
 
-		public MissionObjectives(World world, MissionObjectivesInfo moInfo)
+		public MissionObjectives(World world, MissionObjectivesInfo info)
 		{
-			info = moInfo;
+			Info = info;
 			Objectives = new ReadOnlyList<MissionObjective>(objectives);
-
-			world.ObserveAfterWinOrLose = !info.EarlyGameOver;
 		}
 
 		public int Add(Player player, string description, ObjectiveType type = ObjectiveType.Primary, bool inhibitAnnouncement = false)
@@ -141,7 +139,7 @@ void CheckIfGameIsOver(Player player)
 
 			var gameOver = players.All(p => p.WinState != WinState.Undefined || !p.HasObjectives);
 			if (gameOver)
-				Game.RunAfterDelay(info.GameOverDelay, () =>
+				Game.RunAfterDelay(Info.GameOverDelay, () =>
 				{
 					player.World.EndGame();
 					player.World.SetPauseState(true);
@@ -154,7 +152,7 @@ public void OnPlayerWon(Player player)
 			var players = player.World.Players.Where(p => !p.NonCombatant);
 			var enemies = players.Where(p => !p.IsAlliedWith(player));
 
-			if (info.Cooperative)
+			if (Info.Cooperative)
 			{
 				WinStateCooperative = WinState.Won;
 				var allies = players.Where(p => p.IsAlliedWith(player));
@@ -167,7 +165,7 @@ public void OnPlayerWon(Player player)
 						p.World.OnPlayerWinStateChanged(p);
 					}
 
-					if (info.EarlyGameOver)
+					if (Info.EarlyGameOver)
 						foreach (var p in enemies)
 							p.PlayerActor.Trait<MissionObjectives>().ForceDefeat(p);
 				}
@@ -177,7 +175,7 @@ public void OnPlayerWon(Player player)
 				player.WinState = WinState.Won;
 				player.World.OnPlayerWinStateChanged(player);
 
-				if (info.EarlyGameOver)
+				if (Info.EarlyGameOver)
 					foreach (var p in enemies)
 						p.PlayerActor.Trait<MissionObjectives>().ForceDefeat(p);
 			}
@@ -190,7 +188,7 @@ public void OnPlayerLost(Player player)
 			var players = player.World.Players.Where(p => !p.NonCombatant);
 			var enemies = players.Where(p => !p.IsAlliedWith(player));
 
-			if (info.Cooperative)
+			if (Info.Cooperative)
 			{
 				WinStateCooperative = WinState.Lost;
 				var allies = players.Where(p => p.IsAlliedWith(player));
@@ -203,7 +201,7 @@ public void OnPlayerLost(Player player)
 						p.World.OnPlayerWinStateChanged(p);
 					}
 
-					if (info.EarlyGameOver)
+					if (Info.EarlyGameOver)
 						foreach (var p in enemies)
 							p.PlayerActor.Trait<MissionObjectives>().ForceDefeat(p);
 				}
@@ -213,7 +211,7 @@ public void OnPlayerLost(Player player)
 				player.WinState = WinState.Lost;
 				player.World.OnPlayerWinStateChanged(player);
 
-				if (info.EarlyGameOver)
+				if (Info.EarlyGameOver)
 					foreach (var p in enemies)
 					{
 						p.WinState = WinState.Won;
diff --git a/OpenRA.Mods.Common/Widgets/Logic/Ingame/LoadIngamePlayerOrObserverUILogic.cs b/OpenRA.Mods.Common/Widgets/Logic/Ingame/LoadIngamePlayerOrObserverUILogic.cs
index f0b65cc..948da83 100644
--- a/OpenRA.Mods.Common/Widgets/Logic/Ingame/LoadIngamePlayerOrObserverUILogic.cs
+++ b/OpenRA.Mods.Common/Widgets/Logic/Ingame/LoadIngamePlayerOrObserverUILogic.cs
@@ -9,6 +9,7 @@
 #endregion
 
 using OpenRA.Mods.Common.Scripting;
+using OpenRA.Mods.Common.Traits;
 using OpenRA.Widgets;
 
 namespace OpenRA.Mods.Common.Widgets.Logic
@@ -29,12 +30,13 @@ public LoadIngamePlayerOrObserverUILogic(Widget widget, World world)
 			{
 				var playerWidgets = Game.LoadWidget(world, "PLAYER_WIDGETS", playerRoot, new WidgetArgs());
 				var sidebarTicker = playerWidgets.Get<LogicTickerWidget>("SIDEBAR_TICKER");
+				var objectives = world.LocalPlayer.PlayerActor.Trait<MissionObjectives>();
 
 				sidebarTicker.OnTick = () =>
 				{
 					// Switch to observer mode after win/loss
-					if (world.ObserveAfterWinOrLose && world.LocalPlayer.WinState != WinState.Undefined)
-						Game.RunAfterTick(() =>
+					if (world.LocalPlayer.WinState != WinState.Undefined)
+						Game.RunAfterDelay(objectives.Info.GameOverDelay, () =>
 						{
 							world.LocalPlayer.Spectating = true;
 							playerRoot.RemoveChildren();
```

pchote commented:
This removes a bunch of duplication that I was never 100% happy about, and fixes a long-standing complaint from many players.  :+1: after addressing @obrakmann's suggestion.

obrakmann committed:
Remove World.ObserveAfterWinOrLose

Fixes RenderPlayer for "All Players" view.

obrakmann committed:
Delay loading the observer UI until the game is actually over

deniz1a commented:
I merged the requested changes and rebased.

obrakmann commented:
:+1: after https://github.com/deniz1a/OpenRA/pull/3 has been merged as well.

obrakmann commented on OpenRA.Game/World.cs:
```diff
@@ -66,11 +66,10 @@ public void EndGame()
 		public Player RenderPlayer
 		{
-			get { return renderPlayer == null || (ObserveAfterWinOrLose && renderPlayer.WinState != WinState.Undefined) ? null : renderPlayer; }
+			get { return renderPlayer == null || renderPlayer.Spectating ? null : renderPlayer; }
```
Ah, sorry, this needs to be `|| renderPlayer.WinState != WinState.Undefined ? null : renderPlayer` after all, otherwise the 'All players' view will break. Please amend the first commit I made with the change.

deniz1a commented:
Should I squash all commits?

obrakmann commented:
I'd leave five commits:

Commit | Included commits
----------|-------------------------
b4532d7 | 498b29c
4cbc4fe | a3ead9f, 756fefd, dd32a4e, 8e1e53d, aa33ddf
d50053f | 
54d21fa | 0604233
d01f55f |

deniz1a commented:
All squishy now.

deniz1a commented:
A small button click for the maintainers, a giant leap for gameplay.

obrakmann commented:
Thanks. [Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/1adbb780808cc4ad810324bf6ed52a32b58e5141%5E...1adbb780808cc4ad810324bf6ed52a32b58e5141)
